text/0000-symbol-name-mangling-v2.md

```diff
@@ -0,0 +1,799 @@
+
+- Once a demangler supports Punycode, it is not much effort to support both encodings. The `u` identifier suffix tells the demangler whether it's Punycode. Otherwise it can just assume UTF-8 which already subsumes ASCII.
+
+### Re-use <disambiguator> for crate disambiguator
+
+The RFC currently proposes to represent crate-ids as an `<identifier>` of the form `<name>_<disambiguator>`. However, the `<identifier>` production already supports disambiguation via its `<disambiguator>` component. The crate disambiguator could be encoded into an disambiguation index.
+
+# Appendix A - Suggested Demangling
+
+This RFC suggests that names are demangled to a form that matches Rust syntax as it is used in source code, compiler error messages and `rustdoc`:
+
+ - Path components should be separated by `::`.
+
+ - If the path root is a `<crate-id>` it should be printed as the crate name. If the context requires it for correctness, the crate disambiguator should be printed too, as in, for example, `std[a0b1c2d3]::collections::HashMap`. In this case `a0b1c2d3` would be the disambiguator. Usually, the disambiguator can be omitted for better readability.
+
+ - If the path root is a trait impl, it should be printed as `<SelfType as Trait>`, like the compiler does in error messages.
+
+ - The list of generic arguments should be demangled as `<T1, T2, T3>`.
+
+ - Identifiers and trait impl path roots can have a numeric disambiguator (the `<disambiguator>` production). The syntactic version of the numeric disambiguator maps to a numeric index. If the disambiguator is not present, this index is 0. If it is of the form `s_` then the index is 1. If it is of the form `s<base-62-digit>_` then the index is `<base-62-digit> + 2`. The suggested demangling of a disambiguator is `[<index>]`. However, for better readability, these disambiguators should usually be omitted in the demangling altogether. Disambiguators with index zero can always be omitted.
+
+ The exception here are closures. Since these do not have a name, the disambiguator is the only thing identifying them. The suggested demangling for closures is thus `{closure}[<index>]`.
+
+ - In a lossless demangling, identifiers from the value namespace should be marked with a `'` suffix in order to avoid conflicts with identifiers from the type namespace. In a user-facing demangling, where such conflicts are acceptable, the suffix can be omitted.
+
+# Appendix B - Interesting Examples
+
+We assume that all examples are defined in a crate named `mycrate[xxx]`.
+
+
+### Free-standing Item
+
+```rust
+mod foo {
+  mod bar {
+    fn baz() {}
+  }
+}
+```
+- unmangled: `mycrate::foo::bar::baz`
+- mangled: `_RN11mycrate_xxx3foo3bar3bazVE`
+
+
+### Item Defined In Inherent Method
+
+```rust
+struct Foo<T>(T);
+
+impl<T> Foo<T> {
+  pub fn bar<U>(_: U) {
+    static QUUX: u32 = 0;
+    // ...
+  }
+}
+```
+- unmangled: `mycrate::Foo::bar::QUUX`
+- mangled: `_RNMN11mycrate_xxx3FooE3barV4QUUXVE`
```

Comment:
Shouldn't these mention type parameters?

Reply:
Sure enough. You want to be able to distinguish between these 2 cases (this code compiles today):

```Rust
struct Foo<U,V>(U,V);

impl<U: Fn()> Foo<U, u32> {
    fn foo() {}
}

impl<U: Fn()> Foo<u32, U> {
    fn foo() {}
}
```

Reply:
OK, we'll have to take care of that then.

Reply:
I had a little chat with @nikomatsakis about this yesterday and the outcome was that:
- we always should encode type parameters in paths like this one and
- we should also always encode parameter bounds in some form because there is no way to find out if they are needed for disambiguation without looking at other impls -- which we want to avoid. The bounds could be encoded in a numeric disambiguator though.

The consequences this has on symbol syntax should be small. We just have to find the best spot for adding parameter bounds.

Reply:
> we should also always encode parameter bounds 

I still think that's not ideal, and I'd prefer having a disambiguated path to the `impl` (which is hidden in the non-verbose mode) and/or to the type parameters.